platforms/windows/hax_wrapper.c

```diff
@@ -31,21 +31,71 @@
 #include "hax_win.h"
 #include "../../core/include/ia32.h"
 
-int max_cpus;
-hax_cpumap_t cpu_online_map;
-
-uint32_t hax_cpuid()
+inline uint32_t hax_cpu_id(void)
 {
-    return KeGetCurrentProcessorNumber();
+    PROCESSOR_NUMBER ProcNumber = {0};
+    return (uint32_t)KeGetCurrentProcessorNumberEx(&ProcNumber);
 }
 
-struct smp_call_parameter
+int cpu_info_init(void)
 {
-    void (*func)(void *);
-    void *param;
-    /* Not used in DPC model*/
-    hax_cpumap_t *cpus;
-};
+    uint32_t size_group, size_pos, count, group, bit;
+
+    memset(&cpu_online_map, 0, sizeof(cpu_online_map));
+
+    cpu_online_map.group_num = KeQueryActiveGroupCount();
+    cpu_online_map.cpu_num = KeQueryActiveProcessorCountEx(ALL_PROCESSOR_GROUPS);
+    if (cpu_online_map.group_num > HAX_MAX_CPU_GROUP ||
+        cpu_online_map.cpu_num > HAX_MAX_CPUS) {
+        hax_log(HAX_LOGE, "Too many cpus %d-%d in system\n",
+                cpu_online_map.cpu_num, cpu_online_map.group_num);
+        return -1;
+    }
+
+    size_group = cpu_online_map.group_num * sizeof(*cpu_online_map.cpu_map);
+    cpu_online_map.cpu_map = hax_vmalloc(size_group, 0);
+    if (!cpu_online_map.cpu_map) {
+        hax_log(HAX_LOGE, "Couldn't allocate cpu_map for cpu_online_map\n");
+        return -1;
+    }
+
+    size_pos = cpu_online_map.cpu_num * sizeof(*cpu_online_map.cpu_pos);
+    cpu_online_map.cpu_pos = hax_vmalloc(size_pos, 0);
+    if (!cpu_online_map.cpu_pos) {
+        hax_log(HAX_LOGE, "Couldn't allocate cpu_pos for cpu_online_map\n");
+        hax_vfree(cpu_online_map.cpu_map, size_group);
+        return -1;
+    }
+
+    count = 0;
+    for (group = 0; group < cpu_online_map.group_num; group++) {
+        cpu_online_map.cpu_map[group].map = (hax_cpumask_t)KeQueryGroupAffinity(
+            group);
```

Comment:
8 spaces indentation for line wrap or align to the parenthesis.

Reply:
Thanks. Updated.